// shannz-playwright/run.js
//CODE BY DAFFA
//Saluran Dia : https://whatsapp.com/channel/0029VaiVeWA8vd1HMUcb6k2S

const axios = require('axios');

const playwright = {
  avLang: ['javascript', 'python', 'java', 'csharp'],
  
  request: async function(language = 'javascript', code) {
    if (!this.avLang.includes(language.toLowerCase())) {
      throw new Error(`Language "${language}" tidak support. Pilih Language yang tersedia: ${this.avLang.join(', ')}`);
    }

    const url = 'https://try.playwright.tech/service/control/run';
    const headers = {
      'authority': 'try.playwright.tech',
      'accept': '*/*',
      'content-type': 'application/json',
      'origin': 'https://try.playwright.tech',
      'referer': 'https://try.playwright.tech/?l=playwright-test',
      'user-agent': 'Postify/1.0.0',
    };

    const data = {
      code: code,
      language: language
    };

    try {
      const response = await axios.post(url, data, { headers });
      const { success, error, version, duration, output, files } = response.data;
      return { 
        success,
        author: 'Shannz',
        playwrightVersion: version,
        result: {
          duration,
          output, 
          files
        }
      }; 
    } catch (error) {
      if (error.response) {
        const { success, error: errMsg, version, duration, output, files } = error.response.data; 
        return {
          success,
          author: 'Shannz',
          playwrightVersion: version,
          result: {
            duration,
            error: errMsg,
            output
          }
        };
      } else {
        throw new Error(error.message);
      }
    }
  }
};

async function run(lang, code) {
  try {
    const result = await playwright.request(lang, code);
    return result;
    console.log(result);
  } catch (error) {
    console.error('Error:', error.message);
  }
}

module.exports = { run };
